ir/src/validation/error.rs

```diff
@@ -185,39 +80,138 @@ impl SemanticError {
-        ident_name: &str,
-        ident_type: &IdentifierType,
-    ) -> Self {
+    // --- INVALID ACCESS ERRORS ------------------------------------------------------------------
+
+    pub(crate) fn invalid_constant_access_type(name: &str, access_type: &AccessType) -> Self {
+        Self::InvalidUsage(format!(
+            "Constant '{name}' cannot be accessed by a {access_type:?}.",
+        ))
+    }
+
+    pub(crate) fn invalid_periodic_column_access_type(name: &str) -> Self {
+        Self::InvalidUsage(format!(
+            "Attempted to access periodic column '{name}' by index, which is not allowed.",
+        ))
+    }
+
+    pub(crate) fn invalid_public_input_access_type(name: &str) -> Self {
+        Self::InvalidUsage(format!(
+            "Public input '{name}' can only be accessed by indexing into the vector.",
+        ))
+    }
+
+    pub(crate) fn invalid_random_value_access_type(name: &str, access_type: &AccessType) -> Self {
+        Self::InvalidUsage(format!(
+            "Random value '{name}' cannot be indexed as a {access_type}.",
+        ))
+    }
+
+    pub(crate) fn invalid_trace_access_type(name: &str, access_type: &AccessType) -> Self {
+        Self::InvalidUsage(format!(
+            "Trace column binding '{name}' cannot be accessed as a {access_type}.",
+        ))
+    }
+
+    pub(crate) fn invalid_variable_access_type(name: &str, access_type: &AccessType) -> Self {
+        Self::InvalidUsage(format!(
+            "Variable '{name}' cannot be accessed as a {access_type}.",
+        ))
+    }
+
+    pub(crate) fn invalid_periodic_column_access_in_bc() -> SemanticError {
+        SemanticError::InvalidUsage(
+            "Periodic columns cannot be used in boundary constraints.".to_string(),
+        )
+    }
+
+    pub(crate) fn invalid_public_input_access_in_ic() -> SemanticError {
+        SemanticError::InvalidUsage(
+            "Public inputs cannot be used in integrity constraints.".to_string(),
+        )
+    }
+
+    pub(crate) fn invalid_random_value_binding_access(ident: &str) -> SemanticError {
         SemanticError::InvalidUsage(format!(
-            "Identifier {ident_name} was declared as a {ident_type} not as a trace column"
+            "Expected {ident} to be a binding to a single random value."
         ))
     }
 
-    pub(crate) fn invalid_trace_binding(ident: &str) -> SemanticError {
+    pub(crate) fn invalid_trace_binding_access(ident: &str) -> SemanticError {
         SemanticError::InvalidUsage(format!(
             "Expected {ident} to be a binding to a single trace column."
         ))
     }
 
+    pub(crate) fn invalid_trace_offset_in_bc(trace_access: &IndexedTraceAccess) -> SemanticError {
+        SemanticError::InvalidUsage(format!(
+            "Attempted to access trace column {} in a boundary constraint with a non-zero row offset of {}.", trace_access.col_idx(), trace_access.row_offset()
+        ))
+    }
+
+    pub(crate) fn vector_access_out_of_bounds(
```

Comment:
optional minor nit: Just wondering if we should abbreviate `out_of_bounds` to `oob` here and elsewhere to keep the names concise if this is a widely recognized abbreviation. Not a strong opinion though (and could be for a future PR).

Reply:
I'd prefer to leave it as it is, since it's more immediately understandable and the purpose of these functions is to improve readability in the rest of the codebase